## Pair Programming - Regresión lineal 7

### ANOVA

---

In [ ]:
# Tratamiento de datos
import pandas as pd
import numpy as np

# Estadísticos
import statsmodels.api as sm
from statsmodels.formula.api import ols

# Configuración warnings
import warnings
warnings.filterwarnings('ignore')

### 1. Abrimos nuestro df.

In [4]:
df_estandarizado= pd.read_csv('../archivos/metro_C.csv', index_col=0)

In [5]:
df_estandarizado.head(2)

index        date      time weekday holiday_cat  traffic_volume  \
0      0  2012-10-02  09:00:00  martes          no            5546   
1      1  2012-10-02  10:00:00  martes          no            4517   

   traffic_box    temp_c   snow_1h  
0   544.504895  0.552310 -0.027233  
1   472.646491  0.637259 -0.027233

### 2. Inicializamos la ANOVA

In [6]:
lm = ols("traffic_box ~ temp_c + weekday + holiday_cat + snow_1h + time", data=df_estandarizado).fit()

In [7]:
sm.stats.anova_lm(lm)

df        sum_sq       mean_sq             F         PR(>F)
weekday          6.0  6.107391e+07  1.017899e+07   2245.281340   0.000000e+00
holiday_cat      1.0  3.015261e+06  3.015261e+06    665.106610  1.136118e-145
time            23.0  1.135052e+09  4.935009e+07  10885.645823   0.000000e+00
temp_c           1.0  2.838679e+06  2.838679e+06    626.156085  2.596410e-137
snow_1h          1.0  2.637730e+04  2.637730e+04      5.818307   1.586392e-02
Residual     48137.0  2.182291e+08  4.533501e+03           NaN            NaN

### 3. Sacamos el summary.

In [8]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            traffic_box   R-squared:                       0.846
Model:                            OLS   Adj. R-squared:                  0.846
Method:                 Least Squares   F-statistic:                     8286.
Date:                Thu, 26 Jan 2023   Prob (F-statistic):               0.00
Time:                        17:52:37   Log-Likelihood:            -2.7111e+05
No. Observations:               48170   AIC:                         5.423e+05
Df Residuals:                   48137   BIC:                         5.426e+05
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept               73.8182      1.692     43.618      0.000      70.501      77.135
weekday[T.jueves]       99.6230      1.152     86.478      0.000      97.365     101.881
weekday[T.lunes]        72.1092      1.142     63.122      0.000      69.870      74.348
weekday[T.martes]       89.9610      1.151     78.176      0.000      87.706      92.216
weekday[T.miercoles]    94.9195      1.148     82.716      0.000      92.670      97.169
weekday[T.sabado]       35.4325      1.151     30.777      0.000      33.176      37.689
weekday[T.viernes]     103.1177      1.151     89.563      0.000     100.861     105.374
holiday_cat[T.si]       -5.8606      8.763     -0.669      0.504     -23.037      11.316
time[T.01:00:00]       -40.0635      2.124    -18.864      0.000     -44.226     -35.901
time[T.02:00:00]       -59.6606      2.131    -27.996      0.000     -63.837     -55.484
time[T.03:00:00]       -61.1649      2.129    -28.724      0.000     -65.339     -56.991
time[T.04:00:00]       -15.3990      2.113     -7.289      0.000     -19.540     -11.258
time[T.05:00:00]       125.3411      2.120     59.130      0.000     121.186     129.496
time[T.06:00:00]       284.4625      2.114    134.576      0.000     280.320     288.606
time[T.07:00:00]       329.6077      2.116    155.782      0.000     325.461     333.755
time[T.08:00:00]       325.4295      2.115    153.832      0.000     321.283     329.576
time[T.09:00:00]       315.2781      2.132    147.908      0.000     311.100     319.456
time[T.10:00:00]       302.3589      2.117    142.832      0.000     298.210     306.508
time[T.11:00:00]       322.6966      2.150    150.076      0.000     318.482     326.911
time[T.12:00:00]       340.7701      2.149    158.537      0.000     336.557     344.983
time[T.13:00:00]       341.0744      2.163    157.660      0.000     336.834     345.315
time[T.14:00:00]       355.2582      2.146    165.581      0.000     351.053     359.463
time[T.15:00:00]       376.8928      2.155    174.857      0.000     372.668     381.117
time[T.16:00:00]       404.8444      2.140    189.189      0.000     400.650     409.039
time[T.17:00:00]       381.5689      2.155    177.059      0.000     377.345     385.793
time[T.18:00:00]       307.0601      2.141    143.397      0.000     302.863     311.257
time[T.19:00:00]       232.8026      2.148    108.375      0.000     228.592     237.013
time[T.20:00:00]       197.0575      2.142     91.996      0.000     192.859     201.256
time[T.21:00:00]       182.8111      2.141     85.379      0.000     178.614     187.008
time[T.22:00:00]       140.6155      2.138     65.774      0.000     136.425     144.806
time[T.23:00:00]        69.8741      2.126     32.869      0.000      65.707      74.041
temp_c                   7.6743      0.307     24.966    

Viendo los resultados concluimos que todas nuestras variables predictoras tienen un valor menor a 0.05, excepto "holiday_cat", y nuestro R-squared tiene un 0.84 de puntuación 😀


### 4. Abrimos nuestro df sin estandarizar.

In [14]:
df_sin_estandarizar= pd.read_csv('../archivos/metro_A.csv', index_col=0)

In [15]:
df_sin_estandarizar.head(2)

date weekday      time holiday_cat  temp_c  snow_1h  traffic_volume
0  2012-10-02  martes  09:00:00          no   15.13      0.0            5545
1  2012-10-02  martes  10:00:00          no   16.21      0.0            4516

In [16]:
lm = ols("traffic_volume ~ temp_c + weekday + holiday_cat + snow_1h + time", data=df_sin_estandarizar).fit()

In [17]:
sm.stats.anova_lm(lm)

df        sum_sq       mean_sq            F         PR(>F)
weekday          6.0  1.020540e+10  1.700900e+09  2536.157366   0.000000e+00
holiday_cat      1.0  4.041737e+08  4.041737e+08   602.650306  2.898363e-132
time            23.0  1.468912e+11  6.386575e+09  9522.814090   0.000000e+00
temp_c           1.0  4.380156e+08  4.380156e+08   653.110888  4.252352e-143
snow_1h          1.0  3.964558e+06  3.964558e+06     5.911424   1.504663e-02
Residual     48154.0  3.229498e+10  6.706604e+05          NaN            NaN

In [18]:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:         traffic_volume   R-squared:                       0.830
Model:                            OLS   Adj. R-squared:                  0.830
Method:                 Least Squares   F-statistic:                     7359.
Date:                Thu, 26 Jan 2023   Prob (F-statistic):               0.00
Time:                        18:00:46   Log-Likelihood:            -3.9160e+05
No. Observations:               48187   AIC:                         7.833e+05
Df Residuals:                   48154   BIC:                         7.835e+05
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
========================================================================================
                           coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------------
Intercept             -116.5344     20.683     -5.634      0.000    -157.073     -75.996
weekday[T.jueves]     1273.5206     14.003     90.949      0.000    1246.075    1300.966
weekday[T.lunes]       925.5646     13.886     66.657      0.000     898.349     952.780
weekday[T.martes]     1155.8521     13.988     82.635      0.000    1128.436    1183.268
weekday[T.miercoles]  1219.0424     13.948     87.396      0.000    1191.703    1246.382
weekday[T.sabado]      411.5146     13.994     29.407      0.000     384.086     438.943
weekday[T.viernes]    1301.0298     13.995     92.965      0.000    1273.600    1328.460
holiday_cat[T.si]     -118.7758    106.587     -1.114      0.265    -327.687      90.135
time[T.01:00:00]      -307.0941     25.831    -11.888      0.000    -357.724    -256.464
time[T.02:00:00]      -446.2546     25.920    -17.217      0.000    -497.058    -395.452
time[T.03:00:00]      -461.5275     25.901    -17.819      0.000    -512.293    -410.762
time[T.04:00:00]      -121.6462     25.700     -4.733      0.000    -172.018     -71.274
time[T.05:00:00]      1280.7297     25.785     49.670      0.000    1230.191    1331.268
time[T.06:00:00]      3320.7675     25.713    129.149      0.000    3270.370    3371.165
time[T.07:00:00]      3923.8657     25.737    152.459      0.000    3873.420    3974.311
time[T.08:00:00]      3767.4474     25.733    146.408      0.000    3717.011    3817.884
time[T.09:00:00]      3554.8576     25.922    137.136      0.000    3504.050    3605.665
time[T.10:00:00]      3344.1085     25.743    129.902      0.000    3293.651    3394.566
time[T.11:00:00]      3617.3079     26.151    138.322      0.000    3566.051    3668.565
time[T.12:00:00]      3862.3144     26.148    147.708      0.000    3811.063    3913.566
time[T.13:00:00]      3862.1333     26.324    146.716      0.000    3810.538    3913.728
time[T.14:00:00]      4059.8514     26.113    155.474      0.000    4008.670    4111.033
time[T.15:00:00]      4372.5661     26.238    166.647      0.000    4321.138    4423.994
time[T.16:00:00]      4797.1244     26.054    184.120      0.000    4746.057    4848.191
time[T.17:00:00]      4450.8983     26.237    169.644      0.000    4399.474    4502.322
time[T.18:00:00]      3386.7638     26.069    129.916      0.000    3335.668    3437.859
time[T.19:00:00]      2412.9482     26.142     92.301      0.000    2361.709    2464.187
time[T.20:00:00]      1979.2648     26.061     75.948      0.000    1928.185    2030.344
time[T.21:00:00]      1820.3833     26.045     69.894      0.000    1769.335    1871.431
time[T.22:00:00]      1358.9583     26.001     52.266      0.000    1307.996    1409.920
time[T.23:00:00]       630.1030     25.854     24.372      0.000     579.429     680.777
temp_c                   7.5954      0.298     25.494    

Concluimos que no hay una gran diferencia en los resultados respecto a nuestro df estandarizado, aunque mínimamente nuestro R2 ha bajado a 0.83 y las variable predictora de "holiday_cat" ha mejorado aunque sigue sin ser menor que 0.05.